bigframes/session.py

```diff
@@ -1449,14 +1449,34 @@ def _start_query(
         sql: str,
         job_config: Optional[bigquery.job.QueryJobConfig] = None,
         max_results: Optional[int] = None,
+        api_methods: Sequence[str] = [],
     ) -> Tuple[bigquery.table.RowIterator, bigquery.QueryJob]:
         """
         Starts query job and waits for results
         """
         if job_config is not None:
-            query_job = self.bqclient.query(sql, job_config=job_config)
+            # If there is no label set
+            if job_config.labels is None:
+                label_values = api_methods
+            else:
+                cur_labels: Sequence[str] = [*job_config.labels.values()]
+                total_label_values = list(cur_labels) + list(api_methods)
+                # If the total number of labels is under the limit of labels count
+                if len(job_config.labels) + len(api_methods) <= _MAX_LABELS_COUNT:
+                    label_values = total_label_values
+                # We capture the latest label if it is out of the length limit of labels count
+                else:
+                    label_values = total_label_values[-_MAX_LABELS_COUNT:]
```

Comment:
We'll need to make sure we save room for `len(job_config.labels)`

Reply:
I recommend saving some of this label logic in it's own function, possibly in https://github.com/googleapis/python-bigquery-dataframes/blob/main/bigframes/core/io.py and writing some unit tests that account for having many existing labels and many API methods.

Reply:
Done refactoring the labels logic in `io.py` file and adding unit test.